ds/ds.go

```diff
@@ -275,6 +276,56 @@ func Lookup(query dsQuery) (string, string, error) {
 	return e.IPs[0].String(), strconv.Itoa(e.Port), nil
 }
 
+// LookupResult is returned over a channel, and has either a result or an error.
+type LookupResult struct {
+	Entry *dnssd.BrowseEntry
+	Error error
+}
+
+// LookupChan returns a chan of *LookupResult and a cancel function.
+// The cancel function can be called when the desired number of results has
+// been received; or for any other reason in fact.
+func LookupChan(query dsQuery) (chan *LookupResult, context.CancelFunc) {
```

Comment:
Take a look at my resubmitted sort patch -- the original intended function was to sort the list of responses before returning.  While this was previously used to just pick the top one (which usually was the one with the lowest load because that was the default sort parameter) -- you probably still want to pick the top n from that list before getting a list of all the possible ones.  Logic could be here in Lookup or in decpu, I'm flexible on that, but I think we need to integrate sorting into whatever solution we end up with.

Reply:
for now, let's start with just matching for now, sorting can come later. 

The go way to do this would be to
func LookupChan(query dsQuery, opts ...func([]*LookupResult) error) {
..
get the host list as []*LookupResult
for o := range opts {
if err := o(l); err != nil {
return err
}
}

And those opts would bring in sorting or whatever else we need. That way, default is not sorted, but sorting can be added.